### PR TITLE
Add card action effects

### PR DESCRIPTION
**Pull request by oliverlee**
TODO
- [x] Fix card drawing
- [x] Rename card resources (worth/wealth -> copper)
- [x] Add arena methods for moving cards, playing action cards (without consuming resources)
- [x] Add action stack
- [ ] ~Add arena method for revealing reactions~

### CODE REVIEW

File: dominion/src/dominion/arena.rs [hunk 1]
```diff
@@ -115,88 +119,103 @@ impl Arena {
...
-        player_id: usize,
-        card: CardKind,
-        location: Location,
-    ) -> Result<()> {
+    pub fn play_action(&mut self, player_id: usize, card: CardKind) -> Result<()> {
         self.check_active(player_id)?;
 
+        let card_index = self.players[player_id]
+            .hand
+            .iter()
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]: This method is doing two things: 
1) find the first index of a card with the given kind in a players' hand.
2) play the gard at the given index.

I can imagine 1 being useful when writing bots but at other times we will already know the index and just need 2.

Comment (mickvangelderen) [reply dominion/src/dominion/arena.rs]: I suggest to put 1 as a convenience method of the player or leave it out of the core library and just take a card hand index in this method.

File: dominion/src/dominion/arena.rs [hunk 2]
```diff
@@ -115,88 +119,103 @@ impl Arena {
...
+            .position(|&x| x == card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        let gains = card.action().ok_or_else(|| Error::InvalidCardChoice)?;
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]: move this into the match, closer to where its used

File: dominion/src/dominion/arena.rs [hunk 3]
```diff
@@ -115,88 +119,103 @@ impl Arena {
...
+                player.play_zone.push(card);
+
+                *total_wealth += gains;
+
+                Ok(())
+            }
+        }
+    }
+
+    pub fn buy_card(&mut self, player_id: usize, card: CardKind) -> Result<()> {
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]: Why would we ever want to call this with a player_id != current player id? It doesn't even work because of the assert in that case. Either remove the assertion or the argument.

Comment (mickvangelderen) [reply dominion/src/dominion/arena.rs]: Removing the assertion means that we'll be using the wrong turn phase. So remove the player_id parameter.

File: dominion/src/dominion/arena.rs [hunk 4]
```diff
@@ -115,88 +119,103 @@ impl Arena {
...
+        self.check_active(player_id)?;
+
+        let mut supply_count = self
+            .supply
+            .get_mut(card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        match &mut self.turn.phase {
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]: This turn phase thing happens often, you can abstract it
```rust
impl TurnPhase {
fn as_buy_phase_mut(&mut self) -> Result<(&mut remaining_buys, &mut total_wealth), Error::WrongTurnPhase> {
match self {
    TurnPhase::Buy { remaining_buys, total_wealth } => Ok((remaining_buys, total_wealth)),
    _ => Err(Error::WrongTurnPhase),
}
}

Instead of the tuple you can have a struct `BuyPhase` with the fields and change the variants to `Buy(BuyPhase)` etc.
```

File: dominion/src/main.rs [hunk 5]
```diff
@@ -56,9 +56,25 @@ fn big_money(arena: &mut Arena, player_id: usize) {
...
+                    total_wealth,
+                } => {
+                    assert!(total_wealth < CardKind::Silver.cost());
+                    Ok(())
+                }
+            }
+        })
+        .unwrap();
 
     arena.end_buy_phase(player_id).unwrap();
```
Comment (mickvangelderen) [inline dominion/src/main.rs]: Since we keep track of the current state (player, turn) and all the other methods check if the game is in the right state, this function can simply be `end_phase()` which instead of erroring just does the right thing?

**Review by mickvangelderen — CHANGES_REQUESTED** [review]
Some suggestions.

File: dominion/src/dominion/arena.rs (hunk 3 above)
Comment (oliverlee) [reply dominion/src/dominion/arena.rs]: When we implement the server/client functionality, it is possible that a player will try and to buy out of turn. We'll need some way to allow some functions (e.g. view the cards in the trash) without other (e.g. buy a card).

If we remove `player_id ` here, we'll need to validate that the client/player is the current one.

File: dominion/src/main.rs (hunk 5 above)
Comment (oliverlee) [reply dominion/src/main.rs]: We would still need to return a result (in case a non-current player) tries to end the current turn phase. I think it would be better to explicitly use `end_action_phase` and `end_buy_phase` in the case that a user sends two `end_phase` commands by mistake somehow.

File: dominion/src/dominion/arena.rs (hunk 1 above)
Comment (oliverlee) [reply dominion/src/dominion/arena.rs]: I think you're right about there being situations where we will already know the index and just need to play the card at the given index but I'll put off changing this until that situation comes up.

File: dominion/src/dominion/arena.rs (hunk 3 above)
Comment (mickvangelderen) [reply dominion/src/dominion/arena.rs]: We'll need to validate a whole bunch of things because we can't trust the input. This does not protect against another player trying to force the current player to buy a curse card. 

The Arena class should not concern itself with privileges, just with ensuring the right state transitions can be made.

File: dominion/src/dominion/arena.rs (hunk 1 above)
Comment (mickvangelderen) [reply dominion/src/dominion/arena.rs]: In all of the test cases you know the card index because you just push it into the hand. 

Lets say you're writing a bot and you want it to play two cards if they're both in the hand. You can't do this with the current API without searching through your hands unnecessarily.

File: dominion/src/dominion/arena.rs (hunk 3 above)
Comment (oliverlee) [reply dominion/src/dominion/arena.rs]: > The Arena class should not concern itself with privileges, just with ensuring the right state transitions can be made.

I agree this makes sense and can see that the function `play_action`/`buy_card`/etc. would be simpler without checking ` player_id`. However, I don't want remove `player_id` as a parameter until the privilege checking is implemented and I want to test out my effect stack idea first =p

Comment (mickvangelderen) [reply dominion/src/dominion/arena.rs]: Then debug_assert!(...) is more appropriate

**Review by mickvangelderen — CHANGES_REQUESTED** [review]

File: dominion/src/dominion/arena.rs [hunk 6]
```diff
@@ -370,154 +457,154 @@ mod tests {
     fn buy_card_not_enough_wealth() {
```
Comment (andebjor) [inline dominion/src/dominion/arena.rs]: Rename to `buy_card_not_enough_copper`?

File: dominion/src/dominion/player.rs [hunk 7]
```diff
@@ -85,44 +86,44 @@ mod tests {
...
         for _ in 0..5 {
             p.discard_pile.push(CardKind::Copper);
         }
 
         p.draw_card();
 
-        assert_eq!(p.deck_pile, vec![CardKind::Copper; 4]);
+        assert_eq!(p.draw_pile, vec![CardKind::Copper; 4]);
         assert_eq!(p.hand, vec![CardKind::Copper]);
     }
```
Comment (andebjor) [inline dominion/src/dominion/player.rs:117]: Suggest adding a test with an empty `draw_pile` and empty `discard_pile`. Should work now, right?

**Review by mickvangelderen — COMMENTED** [review]
Just as a remark, pub(crate) fields are useful if there are no invariants to uphold but you don't want to make the field part of the external API. If there's no invariants to uphold, the field might as well be pub. If there are invariants to uphold, perhaps it is better to expose pub(crate) methods that will do so. I am of the opinion that  they shouldn't be pub(crate) out of lazyness, even in early development, because it creates more opportunities to break invariants. If a bunch of types work closely together then it is reasonable to put the in the same module so private fields can be accessed.

If you have anything specific you'd like me to review let me know, otherwise I will simply approve the changes for the sake of moving forwards.

**Comment by oliverlee** [comment]
👍 We will probably want to restructure the code and module privacy at some
point as most things should be hidden from the public API (I'm guessing
that we'll probably only want arena methods and card types visible).

**Review by mickvangelderen — DISMISSED** [review]
Lets move forward

**Comment by oliverlee** [comment]
I'd like to clean up the commit messages before merging this.

**Comment by oliverlee** [comment]
Although it shouldn't stop anyone from branching off of this commit.
